starctl

```diff
@@ -1,177 +1,27 @@
-eval declare -a JAVA_OPTS=("$JAVA_OPTS")
-
-JAVA_OPTS+=(-Dstargate.libdir="$STARGATE_HOME")
-JAVA_OPTS+=(-Djava.awt.headless=true)
-
-declare -a STARGATE_ARGS=("$@")
-
-# Use either command line args or environment variables, don't mix and match or
-# else we could end up with unexpected config
-if [[ ${#STARGATE_ARGS[@]} -eq 0 ]]; then
-    echo "Using environment for config"
-    if [[ ! -z "$CLUSTER_NAME" ]]; then
-        STARGATE_ARGS+=("--cluster-name" "$CLUSTER_NAME")
-    fi
-
-    if [[ ! -z "$CLUSTER_VERSION" ]]; then
-        STARGATE_ARGS+=("--cluster-version" "$CLUSTER_VERSION")
-    fi
-
-    if [[ ! -z "$SEED" ]]; then
-        STARGATE_ARGS+=("--cluster-seed" "$SEED")
-    fi
-
-    if [[ ! -z "$LISTEN" ]]; then
-        STARGATE_ARGS+=("--listen" "$LISTEN")
-    else
-        LISTEN="$(_ip_address)"
-        STARGATE_ARGS+=("--listen" "$LISTEN")
-    fi
+for sig in INT QUIT HUP TERM EXIT; do
+  trap cleanup "$sig"
+done
 
-    if [[ ! -z "$SEED_PORT" ]]; then
-        STARGATE_ARGS+=("--seed-port" "$SEED_PORT")
-    fi
+cd -P -- "$(dirname -- "$0")" # switch to this dir
+echo $PWD
 
-    if [[ ! -z "$DATACENTER_NAME" ]]; then
-        STARGATE_ARGS+=("--dc" "$DATACENTER_NAME")
-    fi
-
-    if [[ ! -z "$RACK_NAME" ]]; then
-        STARGATE_ARGS+=("--rack" "$RACK_NAME")
-    fi
-
-    if [[ ! -z "$SIMPLE_SNITCH" ]]; then
-        STARGATE_ARGS+=("--simple-snitch")
-    fi
-
-    if [[ ! -z "$DSE" ]]; then
-        STARGATE_ARGS+=("--dse")
-    fi
-
-    if [[ ! -z "$CQL_PORT" ]]; then
-        STARGATE_ARGS+=("--cql-port" "$CQL_PORT")
-    fi
-
-    if [[ ! -z "$ENABLE_AUTH" ]]; then
-        STARGATE_ARGS+=("--enable-auth")
-    fi
-
-    if [[ ! -z "$USE_PROXY_PROTOCOL" ]]; then
-        STARGATE_ARGS+=("--use-proxy-protocol")
-    fi
-
-    if [[ ! -z "$PROXY_DNS_NAME" ]]; then
-        STARGATE_ARGS+=("--proxy-dns-name" "$PROXY_DNS_NAME")
-    fi
-
-    if [[ ! -z "$PROXY_PORT" ]]; then
-        STARGATE_ARGS+=("--proxy-port" "$PROXY_PORT")
-    fi
-
-    if [[ "$DEVELOPER_MODE" = true ]] || [[ "$DEVELOPER_MODE" = 1 ]]; then
-        STARGATE_ARGS+=("--developer-mode")
-    fi
-
-    if [[ ! -z "$EMULATE_DBAAS_DEFAULTS" ]]; then
-        STARGATE_ARGS+=("--emulate-dbaas-defaults")
-    fi
-
-    if [[ ! -z "$BIND_TO_LISTEN_ADDRESS" ]]; then
-        STARGATE_ARGS+=("--bind-to-listen-address")
-    fi
-
-    if [[ ! -z "$JMX_PORT" ]]; then
-        STARGATE_ARGS+=("--jmx-port" "$JMX_PORT")
-    fi
-
-    if [[ ! -z "$DISABLE_DYNAMIC_SNITCH" ]]; then
-        STARGATE_ARGS+=("--disable-dynamic-snitch")
-    fi
-
-    if [[ ! -z "$DISABLE_MBEAN_REGISTRATION" ]]; then
-        STARGATE_ARGS+=("--disable-mbean-registration")
-    fi
+if [ -z $BRIDGE_TOKEN ]; then
+  export BRIDGE_TOKEN=mockAdminToken
+fi
 
-    if [[ ! -z "$DISABLE_BUNDLES_WATCH" ]]; then
-        STARGATE_ARGS+=("--disable-bundles-watch")
-    fi
+# start REST API service
```

Comment:
Would it make sense to start by starting coordinator, before APIs/Services? Since latter require former.
Probably wouldn't make much sense as both take time to spin up but could reduce time for APIs need to wait for coordinator.

Reply:
On the one hand: The API services should be robust enough to handle things correctly if the coordinator is not present (i.e return 503 if no coordinator available?), which is why I didn't bother to try to put in logic for this. 

On the other hand: might be good to fork `starctl-coordinator` first with a short delay to make sure itstarts up correctly. Traditionally the `starctl` (now `starctl-coordinator`) has required parameters. If you just type `./starctl` and don't provide any parameters, it will exit with an error message but the REST service is started anyway. In this case you wouldn't be left with an orphan process since `starctl` does a `wait` on the forked processes and kills them if it is killed, but the error message from `starctl-coordinator` could be obscured.

Reply:
Yes, I 100% agree that Services need to be robust enough to handle outages, including start up delay.

Also seems like configurable optional delay(s) would be a nice (future?) improvement.

Good point about challenges wrt piping of failures -- would be also great, if possible, to catch some sanity-check style startup failure cases wrt coordinator.